pkg/service/catalog.go

```diff
@@ -73,12 +73,13 @@ type service struct {
 // instantiated and passed to the NewPlan() constructor function which will
 // carry out all necessary initialization.
 type PlanProperties struct {
-	ID          string                 `json:"id"`
-	Name        string                 `json:"name"`
-	Description string                 `json:"description"`
-	Free        bool                   `json:"free"`
-	Metadata    *ServicePlanMetadata   `json:"metadata,omitempty"`
-	Extended    map[string]interface{} `json:"-"`
+	ID               string                 `json:"id"`
+	Name             string                 `json:"name"`
+	Description      string                 `json:"description"`
+	Free             bool                   `json:"free"`
+	Metadata         *ServicePlanMetadata   `json:"metadata,omitempty"`
+	Extended         map[string]interface{} `json:"-"`
+	ParameterSchemas *ParameterSchemas      `json:"schemas,omitempty"`
```

Comment:
From OSB client.

Reply:
There's an opportunity to simplify things here. First, I should articulate why `PlanProperties` is a separate thing from a `plan`. The short answer is simply that there was need for `ServiceProperties` to be a separate thing from a `service` and I wanted to keep things consistent...

The more complete answer thus requires us to understand why `ServiceProperies` is a separate thing from a `service`. I did not want module implementors directly instantiating `service` objects. This is why it's unexported from the `service` package, has a corresponding interface `Service` (which _is_ exported) and a constructor-like function `NewService(...)`.

Falling farther down the rabbit hole... Why did I not want module implementors instantiating `service` objects directly? Well... there are various things that I wanted to guarantee would happen automatically when a new `service` object is initialized-- namely I wanted to index plans by ID.

So-- coming back to `plan` and `PlanProperties`. The division (currently) exists only to mirror the division between `service` and `ServiceProperties`, _however_, the fact that this division exists means we can take advantage of the opportunity to do things in the constructor-like function `NewPlan(...)`.

What I'd like to propose is that the right properties for the `PlanProperties` type to expose here are `ProvisioningParametersSchema`, `UpdatingParametersSchema`, and `BindingParametersSchema`. (It should be valid for some or all of these to be `nil`.) The constructor-like function `NewPlan(...)` can then take accountability for wrapping these in the correct types like `instanceSchemas`, `bindingSchemas`, and the `parameterSchemas` type that wraps both of those. Note that the above three, then, do not even need to be exported. This reduces the surface area of the SDK (`service` package) that we expect module implementors to use.

The constructor-like function `NewPlan(...)` can also be a good place to supplement module-implementor-provided schema data with common things like `"http://json-schema.org/draft-04/schema#"` so that module-implementors do not need to concern themselves with inconsequential details.